draft-ietf-httpbis-priority.md

```diff
@@ -108,35 +110,34 @@ The term control stream is used to describe the HTTP/2 stream with identifier
 0x0, and HTTP/3 control stream; see {{Section 6.2.1 of !HTTP3=I-D.ietf-quic-http}}.
 
 
-# Motivation for Replacing HTTP/2 Priorities {#motivation}
+# Motivation for Replacing RFC 7540 Priorities {#motivation}
 
 An important feature of any implementation of a protocol that provides
 multiplexing is the ability to prioritize the sending of information. This was
-an important realization in the design of HTTP/2. Prioritization is a
-difficult problem, so it will always be suboptimal, particularly if one endpoint
-operates in ignorance of the needs of its peer.
-
-HTTP/2 introduced a complex prioritization scheme that uses a combination of
-stream dependencies and weights to describe an unbalanced tree. This scheme has
-suffered from poor deployment and interoperability.
-
-Clients build an HTTP/2 prioritization tree through a series of individual
-stream relationships, which are transferred to the server using HTTP/2 priority
-signals in either of three forms. First, a HEADERS frame with the PRIORITY flag
-set is an explicit signal that includes an Exclusive flag, Stream Dependency
-field, and Weight field. Second, a HEADERS frame with no PRIORITY flag is an
-implicit signal to use the default priority. Third, the PRIORITY frame, which is
-always explicit since it always contains an Exclusive flag, Stream Dependency
-field, and Weight field.
-
-The rich flexibility of tree building is rarely exercised. Experience has shown
-that clients tend to choose a single model optimized for a web use case and
-experiment within the model constraints, or do nothing at all. Furthermore, many
-clients build their prioritization tree in a unique way, which makes it
-difficult for servers to understand their intent and act or intervene
-accordingly.
-
-Many HTTP/2 server implementations do not include support for the priority
+an important realization in the design of HTTP/2 leading to RFC 7540.
+Prioritization is a difficult problem, so it will always be suboptimal,
+particularly if one endpoint operates in ignorance of the needs of its peer.
+
+RFC 7540 {{?RFC7540}} introduced a complex prioritization scheme that uses a
```

Comment:
```suggestion
RFC 7540 {{?RFC7540}} introduced a complex prioritization signaling scheme that uses a
```

Reply:
Glad you picked up on this. There's a stylistic difference between the two docs and I purposefully tried to stick to this docs style. That's to say that here we've tried to refer to the abstract notion of weights and dependencies as a "scheme", and the signals as just bits on the wire. Signals are input to a concrete instance of the scheme, which itself is context or input for making resuours usage decisions (  along with other stuff outside the scheme). For my money, the signalling wasn't the hard part, the other stuff was.

You'll probably have a good retort to this style :)

Reply:
I'll pile in and say that an actual implementation might have a very simple or very complex internal representation of how priority is allocated.  The weights and dependencies modify a conceptual graph (hopefully consistent between the two endpoints), and that graph is condensed into whatever the implementation actually uses internally.

I tend to think that the weights and dependencies are signaling, because they don't actually specify internals of how an implementation must represent priority; they're inputs.  However, an implementation *could* use them directly as its internal representation, in which case they would also be the priority scheme itself.

Reply:
for clarity: it sounds like you're siding with MT, is that right @MikeBishop ?

Reply:
Slightly.  However, I think the reality is that we're not disagreeing that there is a split, but rather which side of the split the shared state sits.

I think the shared state is part of the signal, because the server need not directly implement priorities by strictly following that system.  (As Cloudflare's own past work demonstrated, what's received from the client is only part of what the server might reasonably consider in its scheme.)  It sounds like you consider the shared state to be a scheme, and the output of that scheme is one input to an implementation.

Both are self-consistent, so long as they're used consistently in the document.

Reply:
thanks.

There's a few interesting things on reflection of this document as a whole

* It says it defines a priority scheme
* about half of the uses of "scheme" describe design elements
* about half the uses of "scheme" describe what RFC 7540 said and some foibles
* RFC 7540 doesn't use the term "priority scheme" at all
  * neither does it use "priority signal" or really explicitly use signal in the context of priorities.
  * Instead,  it just says "stream prioritization" and talks very specifically in those terms
  
I'm starting to think this is what is causing a bit of friction, especially as we try to loop in http2bis. Now is a good time to shore up things.

Since http2bis is leaving stream weights and deps behind to focus on mandatory frame formats (signals), I think a path forward could be to rephrase things as "RFC 7540 priorities" and "HTTP/2 priority signals". But continue to use "priority scheme" as a term of art for this thing we are specifying here.

Reply:
To me, a "scheme" has always been more specific (i.e., Round-Robin vs FIFO vs what firefox H2 does) and I've always referred to H2's setup as a "prioritization **system**" in my writings. 

Mapping that to the above, I tend to side more with Martin and Mike that "signal" is a better denominator here, but that's not a hill I'm willing to die on.